### ISSUE TITLE
Does not work sbt 1.4 remote cache feature

### ISSUE THREAD

**Issue by xuwei-k** [playframework/playframework#10490]
### Play Version

2.8.2

### API

Maybe both(Java and Scala)

### Operating System

Mac and Linux (also maybe Windows)

### JDK

1.8

### Library Dependencies

unrelated

### Expected Behavior

it should work sbt 1.4 remote cache feature

### Actual Behavior

recompile even if there is cache

### Reproducible Test Case

- https://github.com/xuwei-k/sbt-remote-cache-playframework/commit/0cc2569f46e50bb0c8dab7c9648a667bc5389116
- https://github.com/xuwei-k/sbt-remote-cache-playframework/runs/1279043026

#### step and log

```
$ sbt pushRemoteCache
```

```
$ sbt clean pullRemoteCache compile
[info] welcome to sbt 1.4.1 (AdoptOpenJDK Java 1.8.0_265)
[info] loading settings for project sbt-remote-cache-playframework-build from plugins.sbt ...
[info] loading project definition from /home/runner/work/sbt-remote-cache-playframework/sbt-remote-cache-playframework/project
[info] loading settings for project sbt-remote-cache-playframework from build.sbt ...
[info] set current project to sbt-remote-cache-playframework (in build file:/home/runner/work/sbt-remote-cache-playframework/sbt-remote-cache-playframework/)
[info] Executing in batch mode. For better performance use sbt's shell
[success] Total time: 0 s, completed Oct 20, 2020 5:01:29 AM
[info] remote cache artifact extracted for Some(cached-compile)
[info] remote cache artifact extracted for Some(cached-test)
[success] Total time: 1 s, completed Oct 20, 2020 5:01:30 AM
[warn] There may be incompatibilities among your library dependencies; run 'evicted' to see detailed eviction warnings.
[info] compiling 5 Scala sources and 1 Java source to /home/runner/work/sbt-remote-cache-playframework/sbt-remote-cache-playframework/target/scala-2.13/classes ...
[info] done compiling
[success] Total time: 9 s, completed Oct 20, 2020 5:01:39 AM
```

#### `app/controllers/Application.scala`

```scala
package controllers

import play.api.mvc._
import javax.inject.Inject

class Application @Inject()(cc: ControllerComponents) extends AbstractController(cc) {

  def index = Action {
    Ok("hello")
  }

}
```

#### `build.sbt`

```scala
scalaVersion := "2.13.3"

libraryDependencies +=  guice

enablePlugins(PlayScala)

pushRemoteCacheTo := Some(
  MavenCache("local-cache", (ThisBuild / baseDirectory).value / "remote-cache")
)

remoteCacheId := "fixed-id"

remoteCacheIdCandidates := Seq(remoteCacheId.value)

pushRemoteCacheConfiguration := pushRemoteCacheConfiguration.value.withOverwrite(true)
```

#### `conf/routes`

```
GET     /                           controllers.Application.index
```

#### `project/build.properties`

```
sbt.version=1.4.1
```

#### `project/plugins.sbt`

```scala
addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.8.2")
```

### Note

playframework routes generator embed current `DATE` in generated source comments.
I think it is the cause this issue 🤔 

- https://github.com/playframework/playframework/blob/a90ab0f0d027bfc81671c329f7e30eb20f7a4cd8/dev-mode/routes-compiler/src/main/twirl/play/routes/compiler/static/routesPrefix.scala.twirl#L3-L6
- https://github.com/playframework/playframework/blob/a90ab0f0d027bfc81671c329f7e30eb20f7a4cd8/dev-mode/routes-compiler/src/main/scala/play/routes/compiler/RoutesGenerator.scala#L55-L56

```
$  git grep @DATE
dev-mode/routes-compiler/src/main/twirl/play/routes/compiler/inject/forwardsRouter.scala.twirl:// @@DATE:@sourceInfo.date
dev-mode/routes-compiler/src/main/twirl/play/routes/compiler/static/javaWrappers.scala.twirl:// @@DATE:@sourceInfo.date
dev-mode/routes-compiler/src/main/twirl/play/routes/compiler/static/javascriptReverseRouter.scala.twirl:// @@DATE:@sourceInfo.date
dev-mode/routes-compiler/src/main/twirl/play/routes/compiler/static/reverseRouter.scala.twirl:// @@DATE:@sourceInfo.date
dev-mode/routes-compiler/src/main/twirl/play/routes/compiler/static/routesPrefix.scala.twirl:// @@DATE:@sourceInfo.date
```

#### `cat target/scala-2.13/routes/main/router/RoutesPrefix.scala`

different **`@DATE`** comment 😢 

##### after first compile

```
// @GENERATOR:play-routes-compiler
// @SOURCE:/home/runner/work/sbt-remote-cache-playframework/sbt-remote-cache-playframework/conf/routes
// @DATE:Tue Oct 20 05:01:03 UTC 2020
```

##### after second compile
(I want to use remote cache but play re-generate different source 😢 )

```
// @GENERATOR:play-routes-compiler
// @SOURCE:/home/runner/work/sbt-remote-cache-playframework/sbt-remote-cache-playframework/conf/routes
// @DATE:Tue Oct 20 05:01:31 UTC 2020
```

Is this a sbt bug or playframework bug? /cc @eed3si9n 
playframework should not embed `@DATE` comment? 🤔

**Comment by xuwei-k** [comment]
There is one other thing routes-generator embed `@SOURCE` as **absolute path**. I think it may become an obstacle between share caches😢

**Comment by raboof** [comment]
> playframework should not embed @DATE comment? :thinking: 

I agree it would be better not to include a date there, to make the output more deterministic

**Comment by xuwei-k** [comment]
https://github.com/playframework/twirl/blob/a769228cd036e2757c3b1ced0ac7dc590d9d5170/compiler/src/main/scala/play/twirl/compiler/TwirlCompiler.scala#L725

Maybe twirl has same issue :cry:

**Comment by xuwei-k** [comment]
workaround:

- https://github.com/xuwei-k/sbt-remote-cache-playframework/commit/0341fb7b967b77401666d1b85a2c609cf7e67be2
- https://github.com/xuwei-k/sbt-remote-cache-playframework/runs/1280612736

```scala
routesGenerator := {
  import play.routes.compiler.{RoutesCompiler, RoutesGenerator, Rule}
  val base = InjectedRoutesGenerator
  new RoutesGenerator {
    override def generate(
      task: RoutesCompiler.RoutesCompilerTask,
      namespace: Option[String],
      rules: List[Rule]
    ): Seq[(String, String)] = {
      val baseDir = (ThisBuild / baseDirectory).value.getCanonicalPath + "/"
      val sourceCommentPrefix = s"// @SOURCE:${baseDir}"
      base.generate(task = task, namespace = namespace, rules = rules).map {
        case (fileName, src) =>
          fileName -> src.linesIterator
            .filterNot(_ contains "@DATE")
            .map {
              case line if line.startsWith(sourceCommentPrefix) =>
                line.replace(baseDir, "")
              case line =>
                line
            }
            .mkString("\n")
      }
    }
    override def id = base.id
  }
}
```

**Comment by ignasi35** [comment]
>  playframework should not embed `@DATE` comment? 🤔

Could this be the `DATE` of the source file (instead of the timestamp when the source code was created?)

**Comment by ignasi35** [comment]
> [playframework/twirl@`a769228`/compiler/src/main/scala/play/twirl/compiler/TwirlCompiler.scala#L725](https://github.com/playframework/twirl/blob/a769228cd036e2757c3b1ced0ac7dc590d9d5170/compiler/src/main/scala/play/twirl/compiler/TwirlCompiler.scala#L725)
> 
> Maybe twirl has same issue 😢

see https://github.com/playframework/twirl/issues/209

**Comment by mkurz** [comment]
For the generated routes files i think we can fix this like we did with twirl by removing the generation date and keeping the source comment relative:
https://github.com/playframework/twirl/pull/378

**Comment by mkurz** [comment]
Fix is here: #10707
